teps/0125-configmaps-and-secrets-as-param-value-source.md

```diff
@@ -0,0 +1,189 @@
+	Value string  `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+type  EnvVarSource  struct {
+	FieldRef *ObjectFieldSelector `json:"fieldRef,omitempty" protobuf:"bytes,1,opt,name=fieldRef"`
+	ResourceFieldRef *ResourceFieldSelector `json:"resourceFieldRef,omitempty" protobuf:"bytes,2,opt,name=resourceFieldRef"`
+	ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
+	SecretKeyRef *SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
+}
+
+```
+
+From context of `PipelineRun` and `TaskRun` params 
+```yaml
+	- FieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference field of the pod.
+	- ResourceFieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference resource of the container.
+	- ConfigMapKeyRef : To reference ConfigMap value
+	- SecretKeyRef : To reference Secret value
+```
+
+### Contract
+
+### API
+
+To support the feature supported outlined above, we also propose a small addition to the API. 
+In `ParamSpec` need to `TaskRef` and `PipelineRef` objects will include a `ValueFrom` field.
+
+```go
+type  Param  struct {
+	Name string  `json:"name"`
+	Value ArrayOrString `json:"value"`
+	// Additional field ValueFrom to fetch value from ConfigMap or Secret
+	ValueFrom *ValueSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+
+// ValueSource represents a source for the value.
+type  ValueSource  struct {
+	ConfigMapKeyRef *corev1.ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
+	SecretKeyRef *corev1.SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
+}
+```
+
+For users trying to refer param in  `PipelineRun` or `TaskRun` the options would look like
+
+```yaml
+spec:
+  params:
+    - name: FROM_CONFIGMAP
+      valueFrom:
+	       configMapKeyRef:
+		       name: pipeline-run-configmap
+		       key: HELLO_WORLD_KEY
+    - name: FROM_SECRET
+      valueFrom:
+	       secretKeyRef:
+		       name: pipeline-run-secret
+		       key: HELLO_WORLD_KEY
+    - name: AS_VALUE
+      value: hello-world 
+```
+
+
+### User Stories (optional)
+
+#### Versioned `Task`s and `Pipeline`s and Pipeline-as-code
+
+
+#### Tooling
+
+
+### Risks and Mitigations
+
+## Test Plan
+
+Will need to test all three possible scenarios
+ - With param value passed from Secret key
+ - With param value passed from Configmap key
+ - With param value passed as value  
+
+## Drawbacks
+
+- Controller will need access to `Secrets` and `Configmaps` across all namespaces
+- 
+## Alternatives
+
+1. TEP 0029-step-workspaces or step env variable
+
+   - pros: not all params values can be passed this way
```

Comment:
Is this an advantage?

Reply:
Removed it.
Just mentioned that it works in script section